### DISCUSSION TITLE
v2 Reporting Refactor

### DISCUSSION THREAD

**Discussion by jwildfire** [Gilead-Public/gsm#1527] [General]
# Overview

Instead of passing complex `assessment` objects to our RMD reports, refactor to send only the required tables, figures and metadata. 

Going through each reporting function below shows that the only components that are really used are the `dfSummary` tables and the `charts`.  

# Data Requirements by Function

The functions to generate a KRI report are ... 

1.` Study_Report()`
2. `rmarkdown::render()`
3. `inst/report/KRIReportBySite.RMD`
4. `MakeReportSetup(assessment, dfSite,strType)` - Various prep work for report. including splitting assessment into `active` and `dropped` objects
   1. `AssessStatus(assessment, strType = strType)` - 
   2. `Overview_Table( lAssessments = output$active,  dfSite = dfSite, strReportType = type)` Requires: **kri$lResults$lData$dfSummary**
   3. `MakeSummaryTable( output$active,   dfSite)`  Requires: **kri$lResults$lData$dfSummary**
   4. `MakeSummaryTable( output$dropped,  dfSite )`
6. `GetSnapshotDate(params$status_study)`
7. `MakeStudyStatusTable(dfStudy = params$status_study, overview_raw_table = setup$overview_raw_table, longitudinal = params$longitudinal)`
8. `Make_Timeline(status_study = params$status_study, longitudinal = params$longitudinal)`
9. MakeOverviewMessage( report = "site",  status_study = params$status_study,  overview_raw_table = setup$overview_raw_table,   red_kris = setup$red_kris )
10. MakeResultsTable( assessment = setup$active,   summary_table = setup$summary_table,   lCharts = params$lCharts  ) Just uses **charts** from assessments
11. MakeKRIGlossary(  strWorkflowIDs = names(setup$active),  lStatus = params$status_snap)
12. MakeErrorLog(setup$active)
13. MakeResultsTable(setup$dropped, setup$dropped_summary_table)

**Comment by william-lee-96** [comment]
Hi all, I've been using the gsm HTML site-level report inside of a demo Shiny app to be deployed to Posit Connect. Everything works great locally, but when I deploy the app to Posit Connect, I get this error related to the `Study_Report()` function:

```
2024/05/29 3:52:43 PM: Warning in file(con, "w") :
2024/05/29 3:52:43 PM:   cannot open file 'KRIReportBySite.knit.md': Read-only file system
2024/05/29 3:52:43 PM: Error in file(con, "w") : cannot open the connection
2024/05/29 3:52:43 PM: Calls: local ... tryCatch -> tryCatchList -> tryCatchOne -> 
2024/05/29 3:52:43 PM: Execution halted
2024/05/29 3:52:43 PM: Shiny application exiting ...
```

Within `Study_Report()`, I know it's this `rmarkdown::render()` call at the end that's generating the intermediate "KRIReportBySite.knit.md" file:

```
  # render
  rmarkdown::render(
    projectTemplate,
    output_file = strOutpath,
    params = list(
      assessment = lAssessments,
      status_study = dfStudy,
      status_site = dfSite,
      status_snap = lStatus,
      longitudinal = lLongitudinal,
      lCharts = lCharts
    ),
    envir = new.env(parent = globalenv())
  )
```

While `Study_Report()` has the "strOutpath" parameter which allows me to change where the actual HTML report gets written to, I don't see an option to change where this intermediate "KRIReportBySite.knit.md" file gets written to (which, currently, appears to be the same directory as the "KRIReportBySite.Rmd" template).

I've attempted various workarounds utilizing the withr package, `setwd()`, and `Sys.setenv()` to get the intermediate file written to a temporary directory, but no luck -- always the same error as above. Do you have a known solution/workaround for this in v1, and if not, is this something that will be addressed in v2 (e.g., by adding a parameter to `Study_Report()` so the user can pass their own path to `rmarkdown::render()`'s "intermediates_dir" parameter? Thank you!

**Comment by jwildfire** [comment]
@william-lee-96 I think you'd just need to call `rmarkdown::render` directly in Shiny and set the parameters. 

`Study_Report` has been [streamlined quite a bit in v2](https://github.com/Gilead-BioStats/gsm/blob/v2-dev/R/Report_KRI.R), so it's basically just becoming a pass-through for the `render()` function. If that doesn't fix things, we could consider exposing `...` in `KRI_Report` and passing it through to `render`, but not sure it's needed with the simplification. 

Happy to chat further and/or have a quick call to help troubleshoot.

**Reply by william-lee-96** [reply]
Thanks for your quick response @jwildfire!  So just to be clear, the best solution here would be to take the code inside of `Study_Report` (up to the `rmarkdown::render` call), wrap it in another function, and then feed the output of that function into my own `rmarkdown::render` call (with the "intermediates_dir" parameter set to my desired path)?

**Reply by jwildfire** [reply]
Yep - I think that's the right path for v1.X. 

Would be good for you to try out v2 once it's a little further along.  Happy to consider updating parameters in v2 once we've got core functionality finalized in a few weeks.

**Reply by william-lee-96** [reply]
Great -- thanks so much for your input, Jeremy! I ended up just wrapping the `Study_Report` code in my own function `Study_Report_WL`, identical to the former except now there is the optional parameter "intermediates_dir" to specify a directory for intermediate files (the argument is passed to `rmarkdown::render`'s "intermediates_dir" parameter). With this workaround, my Shiny app is now successfully deploying to Posit Connect.

And yes, would love to try out v2 whenever it's ready!

**Comment by william-lee-96** [comment]
Hi @jwildfire, I wanted to ask a few gsm-related questions that came out of our CSM group:

Regarding the 12 standard KRIs in the HTML report:
- In slide 2 of this [presentation](https://drive.google.com/file/d/1Dlu8CoGF4TBCQwOQM75VJtGqTohsQ15F/view), we can see that for KRI0001, the "Analysis Set" is filtered to just enrolled subjects. Is there documentation on which "Analysis Set" is being used for each KRI?
- Secondarily, are there plans to add an "Analysis Set" column in the HTML report KRI Glossary (where the value for KRI0001 would be something like "Enrolled") in v2? This is something we'd be very interested in!
- `Days on Study` is defined as "Enroll to d/c date" -- does "d/c date" mean "discontinuation/completion date"?

And regarding what gets sent to the TMF:
- Are you just sending the HTML report as is?

Thank you!
